File: android/app/src/main/java/app/selfagent/ledger/PayParser.kt
package app.selfagent.ledger

data class PendingTxn(
    val id: String,
    val amount: Double?,
    val dir: String,
    val title: String,
    val source: String,
    val accountHint: String,
    val category: String,
    val raw: String,
    val at: Long,
    val channel: String = "unknown",
)

object PayParser {
    private val payHint = Regex("支付成功|付款成功|已支付|已付款|消费成功|转账成功|收款成功|到账|收款")
    private val yen = Regex("[¥￥]\\s*(\\d+(?:\\.\\d{1,2})?)")
    private val yuan = Regex("(\\d+(?:\\.\\d{1,2})?)\\s*元")

    fun sourceOf(pkg: String) = when (pkg) {
        "com.tencent.mm" -> "wechat"
        "com.eg.android.AlipayGphone" -> "alipay"
        "com.unionpay" -> "unionpay"
        else -> "other"
    }

    fun isPayText(raw: String) = payHint.containsMatchIn(raw)

    fun parseAmount(raw: String): Double? =
        yen.find(raw)?.groupValues?.get(1)?.toDoubleOrNull()
            ?: yuan.find(raw)?.groupValues?.get(1)?.toDoubleOrNull()

    fun parse(
        pkg: String,
        raw: String,
        at: Long = System.currentTimeMillis(),
        channel: String = "unknown",
        sourceEventId: String? = null,
    ): PendingTxn? {
        val text = raw.replace(Regex("\\s+"), " ").trim()
        if (text.isEmpty() || !isPayText(text)) return null
        val source = sourceOf(pkg)
        val amount = parseAmount(text)
        val incoming = Regex("收款|到账").containsMatchIn(text) &&
            !Regex("支付成功|付款成功|已支付|消费成功").containsMatchIn(text)
        val merchant = Regex("(?:向|给)\\s*([^\\s，。,¥￥]{2,16})").find(text)
            ?.groupValues?.get(1)?.replace(Regex("支付|付款"), "")?.trim()
            ?: Regex("(?:来自|商户|商家)[:：]?\\s*([^\\s，。,¥￥]{2,16})").find(text)
                ?.groupValues?.get(1)?.trim()
        val title = merchant?.ifBlank { null }
            ?: if (source == "wechat") "微信支付" else if (source == "alipay") "支付宝" else if (source == "unionpay") "云闪付" else "支付"
        val category = when {
            incoming -> "收入"
            Regex("餐|饭|面|外卖|咖啡|奶茶|星巴克").containsMatchIn(text) -> "餐饮"
            Regex("地铁|打车|滴滴|公交|加油").containsMatchIn(text) -> "交通"
            else -> "其他"
        }
        val dir = if (incoming) "in" else "out"
        val id = sourceEventId?.trim()?.takeIf { it.isNotEmpty() }
            ?: java.util.UUID.nameUUIDFromBytes(
                listOf(source, amount?.toString() ?: "na", title, dir, channel, at.toString()).joinToString("|").toByteArray()
            ).toString()
        return PendingTxn(
            id = id,
            amount = amount,
            dir = dir,
            title = title,
            source = source,
            accountHint = if (source == "alipay") "支付宝" else if (source == "wechat") "微信零钱/银行卡" else "资金账户",
            category = category,
            raw = text,
            at = at,
            channel = channel,
        )
    }
}

File: _quarantine/old-frontend-src/lib/pay-parser.ts
import type { PaySource, PendingTxn } from "./finance-types";

export const PAY_PACKAGES: Record<string, PaySource> = {
  "com.tencent.mm": "wechat",
  "com.eg.android.AlipayGphone": "alipay",
  "com.unionpay": "unionpay",
};

const PAY_HINT = /支付成功|付款成功|已支付|已付款|消费成功|转账成功|收款成功|到账|收款/;

export const DEMO_NOTICES: { pkg: string; raw: string }[] = [
  { pkg: "com.eg.android.AlipayGphone", raw: "支付宝 支付成功 向星巴克支付 ¥36.00" },
  { pkg: "com.tencent.mm", raw: "微信支付 支付成功 已付款" },
  { pkg: "com.eg.android.AlipayGphone", raw: "支付宝 收款到账 ¥200.00 来自张三" },
  { pkg: "com.unionpay", raw: "云闪付 消费成功 ¥18.50 地铁" },
];

export function isPayText(raw: string) {
  return PAY_HINT.test(raw);
}

export function sourceOf(pkg: string): PaySource {
  return PAY_PACKAGES[pkg] ?? "other";
}

export function parseAmount(raw: string): number | null {
  const yen = raw.match(/[¥￥]\s*(\d+(?:\.\d{1,2})?)/);
  if (yen) return Number(yen[1]);
  const yuan = raw.match(/(\d+(?:\.\d{1,2})?)\s*元/);
  if (yuan) return Number(yuan[1]);
  return null;
}

export function guessMerchant(raw: string): string | null {
  const toward = raw.match(/(?:向|给)\s*([^\s，。,¥￥支付付款成功]{2,16})/);
  if (toward) return toward[1].replace(/支付|付款/g, "").trim() || null;
  const from = raw.match(/(?:来自|商户|商家)[:：]?\s*([^\s，。,¥￥]{2,16})/);
  if (from) return from[1].trim();
  if (/外卖/.test(raw)) return "外卖";
  if (/地铁/.test(raw)) return "地铁";
  return null;
}

export function guessCategory(text: string) {
  if (/餐|饭|面|外卖|咖啡|奶茶|星巴克/.test(text)) return "餐饮";
  if (/地铁|打车|滴滴|公交|加油|交通/.test(text)) return "交通";
  if (/超市|购物|淘宝|京东|拼多多/.test(text)) return "购物";
  if (/会员|订阅|视频/.test(text)) return "订阅";
  return "其他";
}

export function parsePayText(pkg: string, raw: string, at = Date.now()): PendingTxn | null {
  const text = raw.replace(/\s+/g, " ").trim();
  if (!text || !isPayText(text)) return null;
  const source = sourceOf(pkg);
  const amount = parseAmount(text);
  const incoming = /收款|到账/.test(text) && !/支付成功|付款成功|已支付|消费成功/.test(text);
  const merchant = guessMerchant(text);
  const title =
    merchant ||
    (source === "wechat" ? "微信支付" : source === "alipay" ? "支付宝" : source === "unionpay" ? "云闪付" : "支付");
  return {
    id: crypto.randomUUID(),
    amount,
    dir: incoming ? "in" : "out",
    title,
    source,
    accountHint: source === "alipay" ? "支付宝" : source === "wechat" ? "微信零钱/银行卡" : "资金账户",
    category: incoming ? "收入" : guessCategory(text + (merchant ?? "")),
    raw: text,
    at,
  };
}

export function isDuplicate(a: PendingTxn, b: PendingTxn) {
  return a.source === b.source && a.amount === b.amount && a.title === b.title && Math.abs(a.at - b.at) < 10_000;
}
